x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -303,97 +304,71 @@ public void analyzedPlan(
+
+            // "reset" execution information for all ccs or non-ccs (local) clusters, since we are performing the indices
+            // resolving one more time (the first attempt failed and the query had a filter)
+            for (String clusterAlias : executionInfo.clusterAliases()) {
+                executionInfo.swapCluster(clusterAlias, (k, v) -> null);
+            }
 
-    private void preAnalyzeLookupIndices(List<TableInfo> indices, ListenerResult listenerResult, ActionListener<ListenerResult> listener) {
-        if (indices.size() > 1) {
-            // Note: JOINs on more than one index are not yet supported
-            listener.onFailure(new MappingException("More than one LOOKUP JOIN is not supported"));
-        } else if (indices.size() == 1) {
-            TableInfo tableInfo = indices.get(0);
-            TableIdentifier table = tableInfo.id();
-            // call the EsqlResolveFieldsAction (field-caps) to resolve indices and get field types
-            indexResolver.resolveAsMergedMapping(
-                table.index(),
-                Set.of("*"), // TODO: for LOOKUP JOIN, this currently declares all lookup index fields relevant and might fetch too many.
-                null,
-                listener.map(indexResolution -> listenerResult.withLookupIndexResolution(indexResolution))
-            );
-            // TODO: Verify that the resolved index actually has indexMode: "lookup"
-        } else {
+            // here the requestFilter is set to null, performing the pre-analysis after the first step failed
+            preAnalyzeIndices(preAnalysis.indices, executionInfo, listenerResult, null, l);
+        }).<LogicalPlan>andThen((l, listenerResult) -> {
+            assert requestFilter != null : "The second analysis shouldn't take place when there is no index filter in the request";
+            LOGGER.debug("Analyzing the plan (second attempt, without filter)");
+            LogicalPlan plan;
             try {
-                // No lookup indices specified
-                listener.onResponse(
-                    new ListenerResult(
-                        listenerResult.indices,
-                        IndexResolution.invalid("[none specified]"),
-                        listenerResult.enrichResolution,
-                        listenerResult.fieldNames
-                    )
-                );
-            } catch (Exception ex) {
-                listener.onFailure(ex);
+                plan = analyzeAction.apply(listenerResult);
+            } catch (Exception e) {
+                l.onFailure(e);
+                return;
             }
-        }
+            LOGGER.debug("Analyzed plan (second attempt, without filter):\n{}", plan);
+            l.onResponse(plan);
+        }).addListener(logicalPlanListener);
+    }
+
+    private void preAnalyzeLookupIndices(TableInfo tableInfo, ListenerResult listenerResult, ActionListener<ListenerResult> listener) {
```

Comment:
nit:
```suggestion
    private void preAnalyzeLookupIndex(TableInfo tableInfo, ListenerResult listenerResult, ActionListener<ListenerResult> listener) {
```

Reply:
Done in a separate commit, since accepting this suggestion here would cause compilation errors.